Use stable topological sort in fuse_by_partitions

SherlockNoMad opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #167397


legalize_graph() performs a topo sort that shuffles the nodes is a global way, making the result unpredictable. 
We should avoid this in graph pass in general. 

This problem is discovered when testing regional_inductor, a single fuse region trigger the global reordering. 

Before
https://www.internalfb.com/intern/diffing/?before_paste_number=2029217728&after_paste_number=2029218006&regex_remove_pattern=&enable_regex_remove=0&strip_empty_lines=0&line_wrap=0&selected_tab=plain_diff

After
https://www.internalfb.com/intern/diffing/?paste_number=2029162294&regex_remove_pattern=&enable_regex_remove=0&strip_empty_lines=0&line_wrap=0&selected_tab=plain_diff

Left is gm before regional_inductor, right is after. 

cc @ezyang @EikanWang @jgong5 @wenzhe-nrv

SherlockNoMad committed:
Avoid topological sort in fuse_by_partitions

[ghstack-poisoned]

eellison reviewed (commented):
If you are doing manipulations by looking at DAG, you need to topo sort after, or do more work to keep it legalized while doing the matching. This is what we discussed in the meeting with @anijain2305.

Maybe `stable_topological_sort` will work better here ?

ezyang commented on torch/fx/passes/utils/fuser_utils.py:
```diff
@@ -220,22 +220,34 @@ def insert_subgm(
+        output_node = sub_gm.graph.output_node()
+
+    next_node = module_node.next
+    with gm.graph.inserting_before(next_node):
```
OK, I sat down and understood the algorithm. The reason I think this doesn't work is because you can be in this situation:

```
input0 = ..
input1 = ...
output0 = ...
not_in_subgraph = f(output0)
input2 = ...
output1 = ...
```

Here, we DO need to do some code motion for the not_in_subgraph node, since it is "inside" the set of nodes we want to fuse and we need to push it to the end. But it's not necessarily obvious if we should push an arbitrary node in the middle of the fusion region to the end or the beginning. If instead it didn't depend on output0 but produced input2, then we would need to move it to the beginning. But I think you can make this work.

SherlockNoMad replied:
indeed, the UT failure caught this case.

I changed the impl to use stable topo sort instead.

ezyang reviewed (changes requested):
algo no work

SherlockNoMad committed:
Update on "Avoid topological sort in fuse_by_partitions"



legalize_graph() performs a topo sort that shuffles the nodes is a global way, making the result unpredictable. 
We should avoid this in graph pass in general. 

This problem is discovered when testing regional_inductor, a single fuse region trigger the global reordering. 

Before
https://www.internalfb.com/intern/diffing/?before_paste_number=2029217728&after_paste_number=2029218006&regex_remove_pattern=&enable_regex_remove=0&strip_empty_lines=0&line_wrap=0&selected_tab=plain_diff

After
https://www.internalfb.com/intern/diffing/?paste_number=2029162294&regex_remove_pattern=&enable_regex_remove=0&strip_empty_lines=0&line_wrap=0&selected_tab=plain_diff

Left is gm before regional_inductor, right is after. 

cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

SherlockNoMad committed:
Update on "Use stable topological sort in fuse_by_partitions"



legalize_graph() performs a topo sort that shuffles the nodes is a global way, making the result unpredictable. 
We should avoid this in graph pass in general. 

This problem is discovered when testing regional_inductor, a single fuse region trigger the global reordering. 

Before
https://www.internalfb.com/intern/diffing/?before_paste_number=2029217728&after_paste_number=2029218006&regex_remove_pattern=&enable_regex_remove=0&strip_empty_lines=0&line_wrap=0&selected_tab=plain_diff

After
https://www.internalfb.com/intern/diffing/?paste_number=2029162294&regex_remove_pattern=&enable_regex_remove=0&strip_empty_lines=0&line_wrap=0&selected_tab=plain_diff

Left is gm before regional_inductor, right is after. 

cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

SherlockNoMad committed:
Update on "Use stable topological sort in fuse_by_partitions"



legalize_graph() performs a topo sort that shuffles the nodes is a global way, making the result unpredictable. 
We should avoid this in graph pass in general. 

This problem is discovered when testing regional_inductor, a single fuse region trigger the global reordering. 

Before
https://www.internalfb.com/intern/diffing/?before_paste_number=2029217728&after_paste_number=2029218006&regex_remove_pattern=&enable_regex_remove=0&strip_empty_lines=0&line_wrap=0&selected_tab=plain_diff

After
https://www.internalfb.com/intern/diffing/?paste_number=2029162294&regex_remove_pattern=&enable_regex_remove=0&strip_empty_lines=0&line_wrap=0&selected_tab=plain_diff

Left is gm before regional_inductor, right is after. 

cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

SherlockNoMad committed:
Update on "Use stable topological sort in fuse_by_partitions"



legalize_graph() performs a topo sort that shuffles the nodes is a global way, making the result unpredictable. 
We should avoid this in graph pass in general. 

This problem is discovered when testing regional_inductor, a single fuse region trigger the global reordering. 

Before
https://www.internalfb.com/intern/diffing/?before_paste_number=2029217728&after_paste_number=2029218006&regex_remove_pattern=&enable_regex_remove=0&strip_empty_lines=0&line_wrap=0&selected_tab=plain_diff

After
https://www.internalfb.com/intern/diffing/?paste_number=2029162294&regex_remove_pattern=&enable_regex_remove=0&strip_empty_lines=0&line_wrap=0&selected_tab=plain_diff

Left is gm before regional_inductor, right is after. 

cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

SherlockNoMad committed:
Update on "Use stable topological sort in fuse_by_partitions"



legalize_graph() performs a topo sort that shuffles the nodes is a global way, making the result unpredictable. 
We should avoid this in graph pass in general. 

This problem is discovered when testing regional_inductor, a single fuse region trigger the global reordering. 

Before
https://www.internalfb.com/intern/diffing/?before_paste_number=2029217728&after_paste_number=2029218006&regex_remove_pattern=&enable_regex_remove=0&strip_empty_lines=0&line_wrap=0&selected_tab=plain_diff

After
https://www.internalfb.com/intern/diffing/?paste_number=2029162294&regex_remove_pattern=&enable_regex_remove=0&strip_empty_lines=0&line_wrap=0&selected_tab=plain_diff

Left is gm before regional_inductor, right is after. 

cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

SherlockNoMad committed:
Update on "Use stable topological sort in fuse_by_partitions"



legalize_graph() performs a topo sort that shuffles the nodes is a global way, making the result unpredictable. 
We should avoid this in graph pass in general. 

This problem is discovered when testing regional_inductor, a single fuse region trigger the global reordering. 

Before
https://www.internalfb.com/intern/diffing/?before_paste_number=2029217728&after_paste_number=2029218006&regex_remove_pattern=&enable_regex_remove=0&strip_empty_lines=0&line_wrap=0&selected_tab=plain_diff

After
https://www.internalfb.com/intern/diffing/?paste_number=2029162294&regex_remove_pattern=&enable_regex_remove=0&strip_empty_lines=0&line_wrap=0&selected_tab=plain_diff

Left is gm before regional_inductor, right is after. 

cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]
